packages/mdl-ripple/mixin.js

```diff
@@ -0,0 +1,313 @@
+import {SequentialAnimation} from './animation';
+
+
+/** @typedef {string} */
+var ElementRef;
+
+
+/**
+ * @enum {string}
+ */
+export const Class = {
+  ROOT: 'md-ripple',
+  BACKGROUND: 'md-ripple__background',
+  BACKGROUND_ACTIVE: 'md-ripple__background-active',
+  BACKGROUND_BOUNDED_ACTIVE_FILL: 'md-ripple__background-bounded-active-fill',
+  FOREGROUND: 'md-ripple__foreground',
+  FOREGROUND_CIRCLE: 'md-ripple__foreground-circle',
+  FOREGROUND_BOUNDED_ACTIVE: 'md-ripple__foreground-bounded-active',
+  FOREGROUND_BOUNDED_CIRCLE_ACTIVE: 'md-ripple__foreground-bounded-circle-active',
+};
+
+
+/**
+ * @enum {ElementRef}
+ */
+export const Identifier = {
+  ROOT: Class.ROOT,
+  SURFACE: 'md-ripple__surface',
+  FOREGROUND: Class.FOREGROUND,
+  FOREGROUND_CIRCLE: Class.FOREGROUND_CIRCLE,
+  BACKGROUND: Class.BACKGROUND
+};
+
+
+/** @const {string} */
+const LOG_DECELERATE = 'cubic-bezier(0.157, 0.72, 0.386, 0.987)';
+
+/** @const {number} */
+const TRANSITION_DELTA = 1e-8;
+
+/** @const {number} */
+const UNBOUNDED_EXPAND_DELAY_MS = 80;
+
+/** @const {number} */
+export const MAX_RIPPLES = 4;
+
+
+export default function MDLRippleMixin(renderer) {
+  Object.assign(this, {
+    initMdlRipple_() {
+      this.renderer_ = bindAll(renderer, this);
+
+      /** @type {boolean} */
+      this.isBounded = true;
+
+      /** @private {!Object} */
+      this.boundingRect_;
+
+      /** @private {number} */
+      this.fgIndex_ = 0;
+
+      /** @private {number} */
+      this.maxRadius_;
+
+      /** @private {boolean} */
+      this.isInit_ = false;
+    },
+
+    maybeInitElements_() {
+      if (this.isInit_) {
+        return;
+      }
+
+      let surface = this.renderer_.get(Identifier.SURFACE);
+      let root = this.renderer_.get(Identifier.ROOT);
+
+      this.boundingRect_ = this.renderer_.getBoundingClientRect(surface);
```

Comment:
I feel like we should be wrapping these reads and writes in an rAF call so that way we're not performing a layout at an inappropriate time. We could have this method return a promise that's resolved when all is initialized.